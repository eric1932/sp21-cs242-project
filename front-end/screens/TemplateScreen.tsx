import * as React from 'react';
import {StyleSheet, TouchableHighlight} from 'react-native';
import {View} from '../components/Themed';
import {getTemplateList} from "../utils/API";
import MyProfileItem from "../components/MyProfileItem";
import {Entypo} from "@expo/vector-icons";

export default function TaskScreen() {
  let [templateList, setTemplateList] = React.useState<Array<string>>([])

  React.useEffect(() => {
    getTemplateList().then((result: Array<string>) => {
      setTemplateList([...templateList, ...result])
    })
  }, [])

  return (
    <View style={styles.container}>
      {templateList.map((eachTemplate: string) => (
        <MyProfileItem key={eachTemplate} name={eachTemplate} value={(
          <TouchableHighlight onPress={() => {
            console.warn('asdf')
          }}>
            <Entypo name="plus" size={24} color="black" />
          </TouchableHighlight>
        )} />
      ))}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    // justifyContent: 'center',
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
  },
  separator: {
    marginVertical: 30,
    height: 1,
    width: '80%',
  },
});
