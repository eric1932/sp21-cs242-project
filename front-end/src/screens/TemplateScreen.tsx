import * as React from 'react';
import {Dispatch} from 'react';
import {Button, StyleSheet, Text, TextInput, TouchableHighlight} from 'react-native';
import {View} from '../components/Themed';
import {getTemplateList} from "../utils/API";
import ListItem from "../components/ListItem";
import {Entypo} from "@expo/vector-icons";
import {getToken} from "../utils/Storage";
import {API_BASE_URL} from "../constants/Networking";

function performInstantiateTask(setSubmitting: Dispatch<boolean>,
                                targetTemplateName: string,
                                period: string,
                                note: string,
                                cookies: string,
                                setErrorMessage: Dispatch<string>,
                                setShowDialog: Dispatch<boolean>): () => void {
  return async () => {
    setSubmitting(true)

    const token = await getToken()
    const myHeaders = new Headers();
    myHeaders.append("token", token === null ? '' : token);

    try {
      const response = await fetch(
        `${API_BASE_URL}/task/add/${targetTemplateName}?`
        + `period=${period}&note=${note}&cookies=${cookies}`, {
          method: 'GET',
          headers: myHeaders,
          redirect: 'follow'
        })
      if (response.status == 404) {
        // fail
        setErrorMessage(await response.text())
      } else {
        // success
        setShowDialog(false)
      }
    } catch (e) {
      // fail
      setErrorMessage(e.toString())
    } finally {
      setSubmitting(false)
    }
  };
}

export default function TaskScreen() {
  const [templateList, setTemplateList] = React.useState<Array<string>>([])
  const [showDialog, setShowDialog] = React.useState(false)
  const [targetTemplateName, setTargetTemplateName] = React.useState('')
  const [submitting, setSubmitting] = React.useState(false)
  const [errorMessage, setErrorMessage] = React.useState('')

  const [period, setPeriod] = React.useState('86400')
  const [note, setNote] = React.useState('')
  const [cookies, setCookies] = React.useState('')

  React.useEffect(() => {
    getTemplateList().then((result: Array<string>) => {
      setTemplateList([...templateList, ...result])
    })
  }, [])

  return (
    <View style={styles.container}>
      {showDialog
        ? (<View style={{width: '60%'}}>
          <Text>Creating Task of {targetTemplateName}</Text>
          <TextInput placeholder={'Period (in secs): default 1 day'} onChangeText={(text) => {
            if (text !== '')
              setPeriod(text)  // TODO number only
          }}/>
          <TextInput placeholder={'Note'} onChangeText={(text) => setNote(text)}/>
          <TextInput placeholder={'Cookies'} onChangeText={(text) => setCookies(text)}/>
          <Button title={'Create'}
                  disabled={submitting}
                  onPress={performInstantiateTask(
                    setSubmitting,
                    targetTemplateName,
                    period,
                    note,
                    cookies,
                    setErrorMessage,
                    setShowDialog)}/>
          <Button disabled={submitting} title={'Back'} onPress={() => {
            setShowDialog(false)
          }}/>
          {errorMessage === '' ? undefined
            : <Text>
              {errorMessage}
            </Text>}
        </View>)
        : templateList.map((eachTemplate: string) => (

          <TouchableHighlight
            key={eachTemplate}
            style={{width: '100%'}}
            underlayColor={'#eee'}
            onPress={() => {
              setTargetTemplateName(eachTemplate)
              setShowDialog(true)
            }}>
            <ListItem name={eachTemplate} value={(
              <Entypo name="plus" size={24} color="black"/>
            )}/>
          </TouchableHighlight>
        ))
      }
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    // justifyContent: 'center',
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
  },
  separator: {
    marginVertical: 30,
    height: 1,
    width: '80%',
  },
});
