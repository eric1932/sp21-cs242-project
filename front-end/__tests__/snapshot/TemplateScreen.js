import * as React from 'react';
import renderer from 'react-test-renderer';
import TemplateScreen from "../../src/screens/TemplateScreen";

it('TemplateScreen', () => {
  const tree = renderer.create(
    <TemplateScreen />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
