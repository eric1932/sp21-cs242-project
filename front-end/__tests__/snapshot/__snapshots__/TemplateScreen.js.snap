// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateScreen 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "alignItems": "center",
        "flex": 1,
      },
    ]
  }
/>
`;
